packages/migrations/src/batched-migrations/batched-migrations-runner.ts

```diff
@@ -0,0 +1,288 @@
+
+    let migration = await selectBatchedMigrationForTimestamp(this.options.project, timestamp);
+
+    if (migration.status === 'succeeded') return;
+
+    // If the migration isn't already in the finalizing state, mark it as such.
+    if (migration.status !== 'finalizing') {
+      migration = await updateBatchedMigrationStatus(migration.id, 'finalizing');
+    }
+
+    const migrationFile = await this.getMigrationForIdentifier(identifier);
+    if (!migrationFile) {
+      throw new Error(`No migration found for identifier ${identifier}`);
+    }
+    const MigrationClass = await this.loadMigrationClass(migrationFile);
+    const migrationInstance = new MigrationClass();
+
+    // TODO: we should still lock this migration. Probably what we want to do
+    // is to have the normal run loop check for a runnable migration *before*
+    // acquiring a lock. If it finds a running one, it should grab and hold a
+    // lock just for that specific migration. This allows the instance that's
+    // performing finalization to run without anyone else trying to grab a global lock.
+    const runner = new BatchedMigrationRunner(migration, migrationInstance);
+    await runner.run();
+
+    migration = await selectBatchedMigrationForTimestamp(this.options.project, timestamp);
+
+    if (migration.status === 'succeeded') return;
+
+    throw new Error(
+      `Expected batched migration with identifier ${identifier} to be marked as 'succeeded', but it is '${migration.status}'.`
+    );
+  }
+
+  start() {
+    if (this.running) {
+      throw new Error('BatchedMigrationsRunner is already running');
+    }
+
+    this.loop();
+  }
+
+  async loop() {
+    this.running = true;
+    while (true) {
+      if (this.abortController.signal.aborted) {
+        this.running = false;
+        return;
+      }
+
+      let didWork = false;
+      try {
+        didWork = await this.maybePerformWork(60 * 1000);
+      } catch (err) {
+        this.emit('error', err);
+      }
+
+      // If we did work, we'll immediately try again since there's probably more
+      // work to be done. If not, we'll sleep for a while - maybe some more work
+      // will become available!
+      if (!didWork) {
+        // We provide the signal here so that we can more quickly stop things
+        // when we're shutting down.
+        try {
+          await sleep(30_000, null, { ref: false, signal: this.abortController.signal });
+        } catch (err) {
+          // We don't care about errors here, they should only ever occur when
+          // the AbortController is aborted. Continue to the next iteration of
+          // the loop so we can shut down.
+          continue;
+        }
+      }
+    }
+  }
+
+  private async getOrStartMigration(): Promise<BatchedMigrationRow | null> {
+    // This query will lock all existing rows in the `batched_migrations` table
+    // to guard against race conditions.
+    //
+    // TODO: should this use `LOCK TABLE ...` instead?
```

Comment:
👀 thoughts? This table will probably always be small enough that we can just use lots of row-level locks. Alternatively, we can just use a named lock here.

Reply:
I changed this to use a named lock in 46af675.

Reply:
Sounds fine to me.